### PR TITLE
Update to use shared secret

### PR DESCRIPTION
**Pull request by jrschumacher**
Uses the crypto API Crypto.getRandomValues() to generate a 128 bit
unique shared secret in hex. Additionally compares with a passed appId
to ensure the shared secret is unique.

Added a timestamp to the iframe payload for future ability to expire
appId within the frame.

### CODE REVIEW

File: src/lib/virtru.js [hunk 1]
```diff
@@ -16,14 +33,10 @@ function replaceAppId(data) {
...
-    const userIdMatch = dataStr.match(new RegExp(userIdRegex, 'i'));
-    if (userIdMatch) {
-      userId = userIdMatch[1];
-    }
-    const delegatedAppId = `${userId}_REDACTED`;
-    localStorage.setItem(delegatedAppId, appId);
-    return dataStr.replace(new RegExp(appIdRegex, 'gi'), `"appId":"${delegatedAppId}"`);
+    const sharedSecret = generateSharedSecret(appId);
+    const ts = new Date().toISOString()
+    localStorage.setItem(sharedSecret, JSON.stringify({ appId, ts }));
```
Comment (jherwitz) [inline src/lib/virtru.js]: Is there any way we can also include info about the parent domain in here to further scope the cred? Like:

```
localStorage.setItem(`${parent.origin}:${sharedSecret}`, JSON.stringify({ appId, ts }));
```

Comment (jrschumacher) [reply src/lib/virtru.js]: @jherwitz could you explain in more detail? You don't want the localStorage to leak to other origins?

Comment (jherwitz) [reply src/lib/virtru.js]: As an extra safeguard. So if someone is able to rip off the secret somehow (i.e., malicious script), they couldn't take it and use it from other domains.

File: src/lib/virtru.js [hunk 2]
```diff
@@ -1,6 +1,23 @@
...
+ */
+function generateSharedSecret(uuid) {
+  let secret;
+  let comparator;
+  if (typeof uuid === 'string') {
+    comparator = uuid.replace('-', '');
+  }
+  while (!secret || (comparator && comparator === secret)) {
+    const uin8 = crypto.getRandomValues(new Uint8Array(16));
+    secret = [].slice.call(uin8).map(number => number.toString(16)).join('');
```
Comment (jherwitz) [inline src/lib/virtru.js]: This LGTM, let's see if we can run it by Ryan Speers tomorrow though.

File: src/lib/virtru.js (hunk 1 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: Added it.

File: src/lib/virtru.js [hunk 3]
```diff
@@ -39,9 +81,25 @@ function injectAppId(header, value) {
...
+      const packet = localStorage.getItem(localStorageKey(headerSharedSecret));
+      try {
+        const { appId, origin, ts, sharedSecret } = JSON.parse(packet);
+        if (
+          // The header shared secret and the packet shared secret should match
+          headerSharedSecret === sharedSecret &&
+          // The origins should match
+          getOrigin() === origin &&
+          // The browser current time should be greater than the timestamp
+          new Date() > Date.parse(ts)
```
Comment (jherwitz) [inline src/lib/virtru.js]: Nit: Maybe `>=` ?

File: src/lib/virtru.js [hunk 4]
```diff
@@ -39,9 +81,25 @@ function injectAppId(header, value) {
   if (header.toLowerCase() === 'authorization') {
     const authHeaderMatch = value.match(new RegExp(authHeaderRegex));
     if (authHeaderMatch) {
-      const delegatedAppId = authHeaderMatch[1];
-      const appId = localStorage.getItem(delegatedAppId);
-      return value.replace(delegatedAppId, appId);
+      const headerSharedSecret = authHeaderMatch[1];
+      // Look up the shared secret
+      const packet = localStorage.getItem(localStorageKey(headerSharedSecret));
```
Comment (jherwitz) [inline src/lib/virtru.js]: I'm wondering what the right thing to do on a miss is. Maybe 'return' a 403?

File: src/lib/virtru.js [hunk 5]
```diff
@@ -1,6 +1,45 @@
...
+ * @param {string} uuid uuid to compare with to ensure uniqueness
+ */
+function generateSharedSecret(uuid) {
+  let secret;
+  let comparator;
+  if (typeof uuid === 'string') {
+    comparator = uuid.replace('-', '');
+  }
+  while (!secret || (comparator && comparator === secret)) {
+    const uin8 = crypto.getRandomValues(new Uint8Array(16));
```
Comment (jherwitz) [inline src/lib/virtru.js]: Let's be sure to circle back with Ryan Speers on this one. Can you add a TODO?

File: src/lib/virtru.js [hunk 6]
```diff
@@ -1,6 +1,45 @@
...
-const authHeaderRegex = '\\[\\["([^"]+)';
+const authHeaderRegex = '\\s*\\[\\s*\\[\\s*"([^"]+)';
+
+/**
+ * Get origin
+ * 
+ * @return origin 
+ */
+function getOrigin() {
+  if (window.location.origin) {
```
Comment (jherwitz) [inline src/lib/virtru.js]: When will this not be set?

File: src/lib/virtru.js [hunk 7]
```diff
@@ -1,6 +1,45 @@
...
+}
+
+/**
+ * Local Storage key
+ * 
+ * @param {string} sharedSecret generate the localstorage key
+ * @returns {string} local storage key
+ */
+function localStorageKey(sharedSecret) {
+  return `${getOrigin()}<>${sharedSecret}`;
```
Comment (jherwitz) [inline src/lib/virtru.js]: Haven't seen a delimiter like this before. Seems to be fine though 🤷‍♂

File: src/lib/virtru.js (hunk 3 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: Good call.

File: src/lib/virtru.js (hunk 4 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: Yea we can do that. We'll want to make sure to add a catch-all in the client libraries to pass it back up to the developer _OR_ to log out the user.

File: src/lib/virtru.js (hunk 6 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: @jherwitz in some browsers its not supported so we fall back to constructing our own.

File: src/lib/virtru.js (hunk 7 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: Well I was looking for something which wouldn't be construed for part of a url.... unless that is something we want to do.

Comment (jrschumacher) [reply src/lib/virtru.js]:
```
 control     = <US-ASCII coded characters 00-1F and 7F hexadecimal>
   space       = <US-ASCII coded character 20 hexadecimal>
   delims      = "<" | ">" | "#" | "%" | <">
```

Comment (jherwitz) [reply src/lib/virtru.js]: I'd do `###` but that's just personal preference.

File: src/lib/virtru.js (hunk 6 above)
Comment (jherwitz) [reply src/lib/virtru.js]: 👍

File: src/lib/virtru.js (hunk 7 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: As long as we both understand I'm good with changing it

File: src/lib/virtru.js (hunk 4 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: Done 
https://github.com/virtru/xdomain/pull/3/files#diff-4c50b6a0281071ca609de570dcc31556R136-R140

File: src/lib/virtru.js (hunk 5 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: Meeting today at 2

File: src/lib/virtru.js [hunk 8]
```diff
@@ -1,6 +1,45 @@
 const appIdRegex = '"[^"]*appId[^"]*"\s*:\s*"([^"]+)"';
-const userIdRegex = '"[^"]*userId[^"]*"\\s*:\\s*"([^"]+)"';
-const authHeaderRegex = '\\[\\["([^"]+)';
+const authHeaderRegex = '\\s*\\[\\s*\\[\\s*"([^"]+)';
```
Comment (robbtraister) [inline src/lib/virtru.js]: why are these strings that are passed to the `RegExp` constructor? why not just `const authHeaderRegex = /\s*.../`? it'd be much easier to read

Comment (jrschumacher) [reply src/lib/virtru.js]: Good call.

File: src/lib/virtru.js [hunk 9]
```diff
@@ -1,6 +1,45 @@
...
+
+/**
+ * Secure Math Random
+ * 
+ * @param {string} uuid uuid to compare with to ensure uniqueness
+ */
+function generateSharedSecret(uuid) {
+  let secret;
+  let comparator;
+  if (typeof uuid === 'string') {
```
Comment (robbtraister) [inline src/lib/virtru.js]: do we care about supporting `String` objects? I know that's effectively deprecated, but technically they still exist.

tbh, I would have written this the same way, but as a review, I have to ask.

File: src/lib/virtru.js (hunk 8 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: @robbtraister done

File: src/lib/virtru.js (hunk 9 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: Good question. I think for this solution it's sufficient. I honestly don't think there will ever be a case where the uuid matches the shared secret. I think the odds will be approx 1 in 17 billion.

File: src/lib/virtru.js [hunk 10]
```diff
@@ -39,9 +81,25 @@ function injectAppId(header, value) {
   if (header.toLowerCase() === 'authorization') {
     const authHeaderMatch = value.match(new RegExp(authHeaderRegex));
```
Comment (robbtraister) [inline src/lib/virtru.js]: you don't need to construct a new regexp here

File: src/lib/virtru.js [hunk 11]
```diff
@@ -16,14 +55,17 @@ function replaceAppId(data) {
   const appIdMatch = dataStr.match(new RegExp(appIdRegex, 'i'));
```
Comment (robbtraister) [inline src/lib/virtru.js]: if you had a trailing `i` to the definition above, you don't need to construct a new regexp here

File: src/lib/virtru.js (hunk 5 above)
Comment (rmspeers) [reply src/lib/virtru.js]: Often you have to worry about browser support on this call -- depending on the range of supported browsers you want (I think you may already have functions in your SDK for this that may be reusable where you support a few different browsers?)

File: src/lib/virtru.js [hunk 12]
```diff
@@ -1,6 +1,45 @@
...
+ * 
+ * @param {string} uuid uuid to compare with to ensure uniqueness
+ */
+function generateSharedSecret(uuid) {
+  let secret;
+  let comparator;
+  if (typeof uuid === 'string') {
+    comparator = uuid.replace('-', '');
+  }
+  while (!secret || (comparator && comparator === secret)) {
```
Comment (rmspeers) [inline src/lib/virtru.js]: What is this for? Just to ensure you don't accidentally generate a UUID secret that is equal to the real appId?

File: src/lib/virtru.js [hunk 13]
```diff
@@ -39,9 +81,25 @@ function injectAppId(header, value) {
...
+      const packet = localStorage.getItem(localStorageKey(headerSharedSecret));
+      try {
+        const { appId, origin, ts, sharedSecret } = JSON.parse(packet);
+        if (
+          // The header shared secret and the packet shared secret should match
+          headerSharedSecret === sharedSecret &&
+          // The origins should match
+          getOrigin() === origin &&
+          // The browser current time should be greater or equal to the timestamp
+          new Date() >= Date.parse(ts)
```
Comment (jherwitz) [inline src/lib/virtru.js]: If this isn't necessary we might as well remove it.

File: src/lib/virtru.js (hunk 10 above)
Comment (jrschumacher) [reply src/lib/virtru.js]: Good call Robb

File: src/lib/virtru.js (hunk 12 above)
Comment (jherwitz) [reply src/lib/virtru.js]: Yes that's the intent. Although now that the lengths are different I guess it won't be an issue?

Comment (jrschumacher) [reply src/lib/virtru.js]: Yes that's correct. Not sure if its necessary.

File: src/lib/virtru.js (hunk 5 above)
Comment (jherwitz) [reply src/lib/virtru.js]: Updated to use same keygen code (cloned) as tdf3-js

File: src/lib/virtru.js (hunk 10 above)
Comment (jherwitz) [reply src/lib/virtru.js]: Updated

File: src/lib/virtru.js (hunk 11 above)
Comment (jherwitz) [reply src/lib/virtru.js]: Updated

**Pull request merged**